### PR TITLE
Do not unbind command line keybindings for "all" context

### LINKED ISSUE #454: Unbind-key -a makes using cmdline, search, and goto-url impossible
[Output of newsboat -v]
newsboat r2.14.1-8-gf03c - https://newsboat.org/
Copyright (C) 2006-2015 Andreas Krennmair
Copyright (C) 2015-2019 Alexander Batischev
Copyright (C) 2006-2017 Newsbeuter contributors
Copyright (C) 2017-2019 Newsboat contributors

Newsboat is free software licensed under the MIT License. (Type `newsboat -vv' to see the full text.)
It bundles JSON for Modern C++ library, licensed under the MIT License: https://github.com/nlohmann/json
It bundles an alphanum algorithm implementation licensed under the MIT license: http://www.davekoelle.com/alphanum.html

newsboat r2.14.1-8-gf03c
System: Linux 4.20.10-arch1-1-ARCH (x86_64)
Compiler: g++ 8.2.1 20181127
ncurses: ncurses 6.1.20180127 (compiled with 6.1)
libcurl: libcurl/7.64.0 OpenSSL/1.1.1a zlib/1.2.11 libidn2/2.1.1 libpsl/0.20.2 (+libidn2/2.1.1) libssh2/1.8.0 nghttp2/1.36.0 (compiled with 7.64.0)
SQLite: 3.27.1 (compiled with 3.27.1)
libxml2: compiled with 2.9.9

[Recreating the Issue]

I edited the config file and unbound all default keybindings with "unbind-key -a"
I then added keybindings for all relevant keys, including ENTER and ESC (For open and quit respectively)
However, upon loading newsboat and starting the cmdline (using ':' ), there is no way to actually send that command to newsboat. Pressing ESC does nothing as well.

[Other Details]
With default keybindings, if I open the command line and enter an invalid command (and use ENTER to send it through), newsboat will tell me it's not a valid command. Upon pressing ESC, I can cancel out of that dialog. However unbinding all keys removes those functionalities, even if I rebind ENTER and ESC back to open and quit respectively.

I have not been able to get it working by tweaking what commands ENTER and ESC are bound to. I also tried tweaking what specific dialogs I unbound all keys from, and I found that only if I unbind keys from "feedlist" can I recreate this issue. Unbinding keys from any of the other dialogs does not recreate the issue. However, even when I specifically bound the ENTER and ESC keys to open/quit in the feedlist dialog, the issue persists.

All of the above issues apply also doing anything in newsboat that requires you to manually type/enter something, such as searching and goto-url.

### PR DESCRIPTION
**Pull request by kmws**
This should fix #454. I don't really have a better idea how to do it, if someone has, please share. Also, it's kind of surprising that `OP_INT_END_CMDLINE` doesn't actually handle this.

### CODE REVIEW

**Comment by Minoru** [comment]
> I don't really have a better idea how to do it, if someone has, please share.

I guess you could create a new map, assign `OP_INT_CANCEL_QNA` and `OP_INT_END_QUESTION` their default values, then copy that map into every context? That'll rid us of the inner loop. I don't see what else can be improved here. You can just `newmap[OP] = oldmap[OP]`; if `OP` doesn't exist in `oldmap`, it'll be default-constructed, which for integers mean it'll be zeroed—and enums are integers; Zero in `Operation` is `OP_NIL`, which is exactly what we need.

I do have a question, though. I admit I don't fully understand `Keymap` yet, maybe you'll shed some light on this. It looks like all internal operations need this kind of treatment, not just `OP_INT_CANCEL_QNA` and `OP_INT_END_QUESTION`. For example, `unbind-all -a` should also disable up/down scrolling in command-line history. Do you have any idea about those?

> Also, it's kind of surprising that OP_INT_END_CMDLINE doesn't actually handle this.

Yeah, that's confusing. I guess it's because of cmdline being implemented in terms of QNA, so QNA gets to handle it first?

**Comment by Minoru** [comment]
OS X failed on Travis CI again; I restarted the builds.

Rustfmt failed because 1.38 apparently formats things a bit differently from 1.37. Not this again… :( If you're using `rustup` or is able to install 1.38 some other way, please run `cargo fmt` to fix this. If not, I can do it after I merge this.

**Comment by Minoru** [comment]
Oh, and one more thing: if you reproduced #454, please remove "to-check" label from the issue.

**Comment by kmws** [comment]
>I guess you could create a new map, assign OP_INT_CANCEL_QNA and OP_INT_END_QUESTION their default values, then copy that map into every context?

Nice idea, thank you.

>It looks like all internal operations need this kind of treatment, not just `OP_INT_CANCEL_QNA` and `OP_INT_END_QUESTION`. For example, `unbind-all -a` should also disable up/down scrolling in command-line history. Do you have any idea about those?

I have considered including all the `INT`ernal commands, but in the end decided to just fix the given case. I'm not sure how `QNA_NEXTHIST` and `QNA_PREVHIST` work (from the code perspective) and how to set them up, because when I try `bind-key % qna-next-history` I get an error during startup about not valid command...

> Yeah, that's confusing. I guess it's because of cmdline being implemented in terms of QNA, so QNA gets to handle it first?

I guess you're right, but I don't really grok the difference between QNA and cmdline. What's the cmdline? :)

**Comment by Minoru** [comment]
> I'm not sure how QNA_NEXTHIST and QNA_PREVHIST work (from the code perspective) and how to set them up, because when I try bind-key % qna-next-history I get an error during startup about not valid command...

I don't think you can set them up with `bind-key`; they're internal, after all :) All you can do is test if they are handled or not; for PREVHIST/NEXTHIST, that means pressing up/down arrow keys when in command line or search. This should read lines from ~/.newsboat/history.cmdline or history.search, and display them—just like history in the shell, or any other readline-enabled program. `unbind-key -a` shouldn't disable that behaviour.

`OP_INT_RESIZE` apparently triggers a re-draw to fix right-aligned strings after terminal resize (199cdf8cfb78104397da5fea57d7b28fd6d1d0ea). But I don't understand where it is generated; I only see code that handles it.

`OP_INT_SET` is used for "set" commands inside macros and commadline, I believe.

`OP_INT_GOTO_URL` is what `goto-url` translates into. No idea why it's a separate operation.

I can file issues about those if you don't want to handle them in this PR, but I think it just amounts to adding more conditions in that `if`, and testing.

> I don't really grok the difference between QNA and cmdline. What's the cmdline? :)

QNA is a generic mechanism to show a string at the bottom of the screen, and take a response from the user. Response is a string. User may quit QNA by pressing Escape. It's implemented inside `FormAction` methods `start_qna`, `finished_qna`, and `get_qna_response`.

cmdline is an instance of QNA. It is started by pressing `:` (which is the default for [`cmdline` op](https://newsboat.org/releases/2.17/docs/newsboat.html#cmdline)), it shows ":" at the bottom of the screen, and user may enter [one of the supported commands](https://newsboat.org/releases/2.17/docs/newsboat.html#_command_line).

Another instance of QNA is search: user presses `/` (which is the default for [`open-search` op](https://newsboat.org/releases/2.17/docs/newsboat.html#open-search)), enters a string to search for, and presses Enter to do the search.

You can find a few more instances by grepping for `start_qna`.

**Comment by Minoru** [comment]
I fixed Rust formatting in master (https://github.com/newsboat/newsboat/commit/402e2d1c4f9d6f20c99b0fee72761133605572a7), so you can just rebase on top of it to make the CI pass.

**Comment by kmws** [comment]
Thank you for the write-up, it's much more clear now. I included the mentioned ops.

>guess you could create a new map, assign OP_INT_CANCEL_QNA and OP_INT_END_QUESTION their default values, then copy that map into every context?

...

>You can just newmap[OP] = oldmap[OP]; if OP doesn't exist in oldmap, it'll be default-constructed, which for integers mean it'll be zeroed—and enums are integers; Zero in Operation is OP_NIL, which is exactly what we need.

Umm, at first I thought that I get it, but not so sure now when I have to write it down...

**Comment by Minoru** [comment]
Never mind; for whatever reason, I thought the op is the map's key, when in fact it's not—a string is a key. Sorry for wasting your time with this.

But I did take another look at the code, and—don't we need to do the same checks for in the "else" branch, where a single context is handled? It looks like the constructor binds all default keys in all possible contexts, so we need to preserve them in all contexts too, right?

**Comment by kmws** [comment]
Yes, you're right. I will correct that in the next iteration. :)

Did you want me to leave *all* internal ops (OP_INT_MIN to OP_INT_MAX) in place, or just the ones that you mentioned?

**Comment by Minoru** [comment]
>Did you want me to leave *all* internal ops (OP_INT_MIN to OP_INT_MAX)
>in place, or just the ones that you mentioned?

Looping from min to max is a great idea, actually! I was wondering if the function can be made future-proof by removing the list from it, and turns out it is indeed possible. Let's go with that.

**Comment by kmws** [comment]
> Looping from min to max is a great idea, actually!

Yeah, it would be great, but it turns out that keys not defined `opdescs` of KeyMap.cpp have `<none>` keybinding by default. So I can't really defend them with tests.

**Comment by Minoru** [comment]
I think you could create two `KeyMaps`, call `unset_all_keys` on one, and then loop from `OP_INT_MIN` to `OP_INT_MAX`, checking that for each operation, the bindings in both `KeyMaps` are the same. A second loop, from `OP_NB_MIN` to `OP_NB_MAX`, would skip ops that are `>= OP_INT_MIN && <= OP_INT_MAX`, and ensure that there are no binding in the modified keymap. Does that make sense?

Also, I managed to get my idea from a previous comment working. Looks like this, and passes all tests from this PR:

```c++
// Prepare a map where only internal ops are bound. These ops can't be
// bound by the user, and they're necessary for Newsboat's internal
// workings.
std::map<std::string, Operation> intOpsOnly;
for (int i = 0; opdescs[i].op != OP_NIL; ++i) {
	if (opdescs[i].flags & KM_INTERNAL) {
		intOpsOnly[opdescs[i].default_key] = opdescs[i].op;
	}
}

if (context == "all") {
	for (unsigned int i = 0; contexts[i] != nullptr; i++) {
		keymap_[contexts[i]] = intOpsOnly;
	}
} else {
	keymap_[context] = std::move(intOpsOnly);
}
```

**Comment by kmws** [comment]
Very helpful, thank you.

Let me know if the newest form is acceptable and I understood you correctly.

File: src/keymap.cpp [hunk 1]
```diff
@@ -476,13 +476,13 @@ void KeyMap::unset_key(const std::string& key, const std::string& context)
...
-		keymap_.clear();
-	} else {
-		auto& key_operation_map = keymap_[context];
-		for (auto& map : key_operation_map) {
-			map.second = OP_NIL;
+		for (unsigned int i = 0; contexts[i] != nullptr; i++) {
+			keymap_[contexts[i]] = internal_ops_only;
 		}
+	} else {
+		keymap_[context] = std::move(internal_ops_only);
```
Comment (Minoru) [inline src/keymap.cpp]: This tries to move out of a `const` object.

**Comment by kmws** [comment]
No worries, I hope I didn't mess something up this time. :)

I don't like the idea of copy-pasting the contexts, but I'm not sure if there's a better way to do it?

**Comment by Minoru** [comment]
I actually think it's *good* for tests to have their separate list of contexts. That way, if we remove a context, the tests will immediately tell us we're breaking something. Alas, adding a new context won't break any tests, but then again, it probably won't break any existing functionality either.

This PR looks good to me now, so I'm merging it. Thank you for putting in the work, @kmws!

**Pull request merged**